apps/token-manager/app/src/App.js

```diff
@@ -86,100 +98,151 @@ class App extends React.PureComponent {
+            </div>
+          }
+          secondary={
+            <Button
+              mode="strong"
+              onClick={this.handleLaunchAssignTokensNoHolder}
+              css={`
+                ${layoutName === 'small'
+                  ? `
+                    width: ${5 * GU}px;
+                    height: ${5 * GU}px;
+                    min-width: 0;
+                    padding: 0;
+                  `
+                  : ''}
+              `}
             >
-              {appStateReady && (
-                <AssignVotePanelContent
-                  opened={sidepanelOpened}
-                  tokenDecimals={numData.tokenDecimals}
-                  tokenDecimalsBase={tokenDecimalsBase}
-                  onUpdateTokens={this.handleUpdateTokens}
-                  getHolderBalance={this.getHolderBalance}
-                  maxAccountTokens={maxAccountTokens}
-                  {...assignTokensConfig}
-                />
-              )}
-            </SidePanel>
-          </IdentityProvider>
-        </div>
-      </Main>
+              {layoutName === 'small' ? <IconPlus /> : 'Add tokens'}
+            </Button>
+          }
+        />
+
+        {appStateReady && holders.length > 0 ? (
+          <Holders
+            holders={holders}
+            groupMode={groupMode}
+            maxAccountTokens={maxAccountTokens}
+            tokenAddress={tokenAddress}
+            tokenDecimalsBase={tokenDecimalsBase}
+            tokenName={tokenName}
+            tokenSupply={tokenSupply}
+            tokenSymbol={tokenSymbol}
+            tokenTransfersEnabled={tokenTransfersEnabled}
+            userAccount={connectedAccount}
+            onAssignTokens={this.handleLaunchAssignTokens}
+            onRemoveTokens={this.handleLaunchRemoveTokens}
+          />
+        ) : (
+          !isSyncing && (
+            <EmptyState onActivate={this.handleLaunchAssignTokensNoHolder} />
+          )
+        )}
+
+        <SidePanel
+          title={
+            assignTokensConfig.mode === 'assign'
+              ? 'Add tokens'
+              : 'Remove tokens'
+          }
+          opened={sidepanelOpened}
+          onClose={this.handleSidepanelClose}
+          onTransitionEnd={this.handleSidepanelTransitionEnd}
+        >
+          {appStateReady && (
+            <AssignVotePanelContent
+              opened={sidepanelOpened}
+              tokenDecimals={numData.tokenDecimals}
+              tokenDecimalsBase={tokenDecimalsBase}
+              onUpdateTokens={this.handleUpdateTokens}
+              getHolderBalance={this.getHolderBalance}
+              maxAccountTokens={maxAccountTokens}
+              {...assignTokensConfig}
+            />
+          )}
+        </SidePanel>
+      </IdentityProvider>
     )
   }
 }
 
 export default () => {
   const { api, appState, connectedAccount, requestMenu } = useAragonApi()
+  const theme = useTheme()
+  const themeMode = useThemeMode()
```

Comment:
I've been maintaining a consistent interface to what hooks expose, when it is a handler prepend it with `handle` -> `handleDoSomething` instead of just exposing the inner `useState` method to update the state var, which we usually prepend with `set`.
`const { set: handleThemeModeChange } = useThemeMode()`
What do you think?

Reply:
I’ve been using the `handleX` less and less since we moved to hooks.

With class components, we couldn’t pass methods declared on the class prototype to event handlers, since we needed `this` to be set properly, so we often had these arrow functions being used as wrappers, and often directly setting the state or other things. It is also quite frequent to see these methods doing several things in reaction to a particular event coming from a children, making the “handle + initiator + event” a good name.

With hooks, I find it to be too different to apply the same convention: the callbacks we create are often focused on are often focused in one task, and reused in different cases (e.g. called from another hook), and calling a function named “handleClick” from something else than a click event doesn’t seem right… so I tend to go for a name that reflect “what it does”, rather than “what it should be called from” now.

Reply:
> With hooks, I find it to be too different to apply the same convention: the callbacks we create are often focused on are often focused in one task

```
const { handleNameChange } = useName()
const handleEffectOnNameChange = () => {
  /* side effect */
  handleNameChange()
}
<Component onChange={handleNameChange} />
<OtherComponentTriggersSideEffect onChange={handleEffectOnNameChange} />
```

Sometimes one function yes but we can implement in a way that does not take away from the `handleX` format no?

> and reused in different cases (e.g. called from another hook), and calling a function named “handleClick” from something else than a click event doesn’t seem right… so I tend to go for a name that reflect “what it does”, rather than “what it should be called from” now

I don't think I'd call a method exported from a hook `handleClick` but I would pass it down to a component on the `onClick` handler

Using a "name that reflects what it does" is exactly what I want to use the `X` for, and just preprend it with `handle` makes it explicit, the method is the handler.

I'd rather we stick to a convention that's widely used throughout the code base instead of switching randomly to another one - again as with all team conventions, I call forth the team so that we can decide as a group cc/ @sohkai @2color

Reply:
I’m not sure I follow :thinking: the `handle` prefix was used to indicate that a function was handling an event, but in your examples it is the other way? e.g. `handleNameChange` is not “handling the name change event”, which is how we usually use these prefixes.

Reply:
Let me take a step backwards then:

```
const useName = () => {
const [name, setName]=useState('')
// either (which would make sense in a hook named more like useNameWithInput or what not
return { handleNameChange: ({ target: value }) => setName(value) }
// or depending on the interface that you may want to provide
return { handleNameChange: setName }
}
```

I do understand there is a difference between the two names and this is definitely an over-simplified example, but the idea is that instead of sending out a `setX` from a hook, we can continue using the `handle` prefix as before to refer to the method that will take action when/if the state var changes.

Reply:
Not quite sure what the React community consensus might be, but for me the distinction for `handleX` is that it's going to handle an event or action from one of its children.

I think `setX` being returned from props is the original intention, and the API we return from a hook doesn't necessarily have to be paired with its internal state.

E.g.

```js
const useName = () => {
const [name, setName]=useState('')
// either (which would make sense in a hook named more like useNameWithInput or what not
return { setNameFromEvent: ({ target: value }) => setName(value) }
// or depending on the interface that you may want to provide
return { setName }
}
```

Reply:
> a hook doesn't necessarily have to be paired with its internal state
Sometimes for sure, when it is a generic hook we expose, but when it is a hook directly related to a component (or component tree) I do think it makes sense to be explicit about the handlers's names.